From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Wed, 12 Jul 2023 21:18:53 +0200
Subject: [PATCH] Modifiable Bukkit enums - Remove hard-coded properties -
 Material - getMaterial, matchMaterial

Moves the implementation of `Material` getters `getMaterial(...)` and `matchMaterial(...)`
out of the `Material` class, and adds their abstract definition to `UnsafeValues` so that they can be
implemented in `CraftMagicNumbers` in the `fiddle-server` module.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index ab83cdfb8bac264a32ca09b753ec55715b6810be..5c002b4a28e57c752dc6e8e10e0e03f3359df2b3 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4636,7 +4636,6 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     public static final String LEGACY_PREFIX = org.fiddlemc.fiddle.material.MaterialLegacyNameChecker.LEGACY_MATERIAL_PREFIX; // Fiddle - modifiable Bukkit enums - Material - remove hard-coded properties - is legacy
 
     private final int id;
-    private static final Map<String, Material> BY_NAME = Maps.newHashMap();
     private final int maxStackHardcoded;
     private final short durabilityHardcoded;
     public final Class<?> data;
@@ -4974,7 +4973,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
      */
     @Nullable
     public static Material getMaterial(@NotNull final String name) {
-        return getMaterial(name, false);
+        return org.fiddlemc.fiddle.material.MaterialNameAndKeyConversion.getMaterial(name); // Fiddle - modifiable Bukkit enums - Material - remove hard-coded properties - getMaterial
     }
 
     /**
@@ -4993,17 +4992,10 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
      * @return Material if found, or null
      */
     @Nullable
-    public static Material getMaterial(@NotNull String name, boolean legacyName) {
-        if (legacyName) {
-            if (!name.startsWith(LEGACY_PREFIX)) {
-                name = LEGACY_PREFIX + name;
-            }
-
-            Material match = BY_NAME.get(name);
-            return Bukkit.getUnsafe().fromLegacy(match);
-        }
-
-        return BY_NAME.get(name);
+    // Fiddle start - modifiable Bukkit enums - Material - remove hard-coded properties - getMaterial
+    public static Material getMaterial(@NotNull final String name, boolean legacyName) {
+        return org.fiddlemc.fiddle.material.MaterialNameAndKeyConversion.getMaterial(name, legacyName);
+        // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - getMaterial
     }
 
     /**
@@ -5018,7 +5010,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
      */
     @Nullable
     public static Material matchMaterial(@NotNull final String name) {
-        return matchMaterial(name, false);
+        return org.fiddlemc.fiddle.material.MaterialNameAndKeyConversion.matchMaterial(name); // Fiddle - modifiable Bukkit enums - Material - remove hard-coded properties - matchMaterial
     }
 
     /**
@@ -5035,23 +5027,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
      */
     @Nullable
     public static Material matchMaterial(@NotNull final String name, boolean legacyName) {
-        Preconditions.checkArgument(name != null, "Name cannot be null");
-
-        String filtered = name;
-        if (filtered.startsWith(NamespacedKey.MINECRAFT + ":")) {
-            filtered = filtered.substring((NamespacedKey.MINECRAFT + ":").length());
-        }
-
-        filtered = filtered.toUpperCase(java.util.Locale.ENGLISH);
-
-        filtered = filtered.replaceAll("\\s+", "_").replaceAll("\\W", "");
-        return getMaterial(filtered, legacyName);
-    }
-
-    static {
-        for (Material material : values()) {
-            BY_NAME.put(material.name(), material);
-        }
+        return org.fiddlemc.fiddle.material.MaterialNameAndKeyConversion.matchMaterial(name, legacyName); // Fiddle - modifiable Bukkit enums - Material - remove hard-coded properties - matchMaterial
     }
 
     /**
diff --git a/src/main/java/org/fiddlemc/fiddle/material/MaterialDefaultImplementations.java b/src/main/java/org/fiddlemc/fiddle/material/MaterialDefaultImplementations.java
index 778f6f89415293f89aa6cbc08af7b6a5c53023f2..09851a9182508fac6b5f6803ae36968e5b40983d 100644
--- a/src/main/java/org/fiddlemc/fiddle/material/MaterialDefaultImplementations.java
+++ b/src/main/java/org/fiddlemc/fiddle/material/MaterialDefaultImplementations.java
@@ -532,6 +532,7 @@ public final class MaterialDefaultImplementations {
         }
     );
 
+    @SuppressWarnings({"unchecked", "rawtypes"})
     private static final CachedMaterialProperty<Class<? extends MaterialData>> materialDataConstructorDeclaringClassCache = new CachedMaterialProperty<>(
         Class[]::new,
         material -> {
diff --git a/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java b/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
index b600f06d62ea4b045590d72626d8029be82e89f2..6a43a90bd4fc533cfc5a738c3b196d72711b171e 100644
--- a/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
+++ b/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
@@ -3,11 +3,15 @@
 package org.fiddlemc.fiddle.material;
 
 import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import org.bukkit.Bukkit;
 import org.bukkit.Material;
 import org.bukkit.NamespacedKey;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 import java.util.Locale;
+import java.util.Map;
 
 /**
  * This class provides a number of implementations for the conversion of names and keys
@@ -28,7 +32,23 @@ public final class MaterialNameAndKeyConversion {
         // Based on the original org.bukkit.Material constructor
         material -> NamespacedKey.minecraft(material.name().toLowerCase(Locale.ROOT))
     );
+    // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - key
+
+    // Fiddle start - modifiable Bukkit enums - Material - remove hard-coded properties - getMaterial
+    private static @Nullable Map<String, Material> cachedMaterialsByName = null;
+
+    private static @NotNull Map<String, Material> getCachedMaterialsByName() {
+        if (cachedMaterialsByName == null) {
+            cachedMaterialsByName = Maps.newHashMap();
+            for (Material material : CachedMaterialProperty.getCachedMaterialValues()) {
+                cachedMaterialsByName.put(material.name(), material);
+            }
+        }
+        return cachedMaterialsByName;
+    }
+    // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - getMaterial
 
+    // Fiddle start - modifiable Bukkit enums - Material - remove hard-coded properties - key
     /**
      * @throws IllegalArgumentException If the given {@code material} is {@linkplain Material#isLegacy() legacy}.
      */
@@ -38,4 +58,51 @@ public final class MaterialNameAndKeyConversion {
     }
     // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - key
 
+    // Fiddle start - modifiable Bukkit enums - Material - remove hard-coded properties - getMaterial
+    public static @Nullable Material getMaterial(@NotNull final String name) {
+        return getMaterial(name, false);
+    }
+
+    public static @Nullable Material getMaterial(@NotNull String name, boolean legacyName) {
+        // Based on the original org.bukkit.Material#getMaterial(String, boolean)
+        if (legacyName) {
+            if (!name.startsWith(MaterialLegacyNameChecker.LEGACY_MATERIAL_PREFIX)) {
+                name = MaterialLegacyNameChecker.LEGACY_MATERIAL_PREFIX + name;
+            }
+
+            Material match = getCachedMaterialsByName().get(name);
+            return Bukkit.getUnsafe().fromLegacy(match, false); // Fiddle - modifiable Bukkit enums - Material - make legacy conversion complete
+        }
+
+        return getCachedMaterialsByName().get(name);
+    }
+    // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - getMaterial
+
+    // Fiddle start - modifiable Bukkit enums - Material - remove hard-coded properties - matchMaterial
+    /**
+     * @throws IllegalArgumentException If the given {@code name} is null.
+     */
+    public static @Nullable Material matchMaterial(@NotNull final String name) {
+        return matchMaterial(name, false);
+    }
+
+    /**
+     * @throws IllegalArgumentException If the given {@code name} is null.
+     */
+    public static @Nullable Material matchMaterial(@NotNull final String name, boolean legacyName) {
+        // Based on the original org.bukkit.Material#matchMaterial(String, boolean)
+        Preconditions.checkArgument(name != null, "Name cannot be null");
+
+        String filtered = name;
+        if (filtered.startsWith(NamespacedKey.MINECRAFT + ":")) {
+            filtered = filtered.substring((NamespacedKey.MINECRAFT + ":").length());
+        }
+
+        filtered = filtered.toUpperCase(java.util.Locale.ENGLISH);
+
+        filtered = filtered.replaceAll("\\s+", "_").replaceAll("\\W", "");
+        return getMaterial(filtered, legacyName);
+    }
+    // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - matchMaterial
+
 }
